go/pkg/client/cas.go

```diff
@@ -1141,56 +1178,67 @@ func (c *Client) FlattenActionOutputs(ctx context.Context, ar *repb.ActionResult
 }
 
 // DownloadDirectory downloads the entire directory of given digest.
-func (c *Client) DownloadDirectory(ctx context.Context, d digest.Digest, execRoot string, cache filemetadata.Cache) (map[string]*TreeOutput, error) {
+// It returns the number of logical and real bytes downloaded, which may be different from sum
+// of sizes of the files due to dedupping and compression.
+func (c *Client) DownloadDirectory(ctx context.Context, d digest.Digest, execRoot string, cache filemetadata.Cache) (map[string]*TreeOutput, *MovedBytesMetadata, error) {
 	dir := &repb.Directory{}
-	if err := c.ReadProto(ctx, d, dir); err != nil {
-		return nil, fmt.Errorf("digest %v cannot be mapped to a directory proto: %v", d, err)
+	stats := &MovedBytesMetadata{}
+
+	protoStats, err := c.ReadProto(ctx, d, dir)
+	sumStats(stats, protoStats)
```

Comment:
Go nit: this should be stats.AddFrom(protostats)
And I think your use-case would improve if it returned a self pointer.